packages/flutter/lib/src/widgets/sensitive_content.dart

```diff
@@ -0,0 +1,225 @@
+
+  final Map<int, ViewContentSensitivityState> _contentSensitivityStates = <int, ViewContentSensitivityState> {};
+  final SensitiveContentService _sensitiveContentService = SensitiveContentService();
+
+  static final SensitiveContentSetting _instance = SensitiveContentSetting._();
+
+  /// Registers a [SensitiveContent] widget that will help determine the
+  /// [ContentSensitivity] level for the Flutter view with ID [viewId].
+  static void register(int viewId, ContentSensitivity desiredSensitivityLevel) {
+     _instance._register(viewId, desiredSensitivityLevel);
+  }
+
+  void _register(int viewId, ContentSensitivity desiredSensitivityLevel) {
+    if (!_contentSensitivityStates.containsKey(viewId)) {
+      _contentSensitivityStates[viewId] = ViewContentSensitivityState();
+    }
+
+    // Update SensitiveContent widget count for those with desiredSensitivityLevel.
+    final ViewContentSensitivityState contentSensitivityStateForView = _contentSensitivityStates[viewId]!;
+    contentSensitivityStateForView.addWidgetWithContentSensitivity(desiredSensitivityLevel);
+
+
+    // If only one SensitiveContent widget in the relevant view sets a
+    // ContentSensitivity level, then we can immediately set
+    // desiredSensitivityLevel for the view.
+    if (contentSensitivityStateForView.getTotalNumberOfWidgets() == 1) {
+      // Set content sensitivity level for view as desiredSensitivityLevel and update stored data.
+      _sensitiveContentService.setContentSensitivity(viewId, desiredSensitivityLevel);
+      contentSensitivityStateForView.currentContentSensitivitySetting = desiredSensitivityLevel;
+      return;
+    }
+
+    // Verify that desiredSensitivityLevel should be set in order for sensitive
+    // content in the view to remain secure.
+    if (!shouldSetContentSensitivity(currentSensitivityLevel: contentSensitivityStateForView.currentContentSensitivitySetting, desiredSensitivityLevel: desiredSensitivityLevel)) {
+      return;
+    }
+
+    // Set content sensitivity level for the view as desiredSensitivityLevel and update stored data.
+    _sensitiveContentService.setContentSensitivity(viewId, desiredSensitivityLevel);
+    contentSensitivityStateForView.currentContentSensitivitySetting = desiredSensitivityLevel;
+  }
+
+  /// Unregisters a [SensitiveContent] widget from the Flutter view with ID
+  /// [viewId].
+  static void unregister(int viewId, ContentSensitivity widgetSensitivityLevel) {
+     _instance._unregister(viewId, widgetSensitivityLevel);
+  }
+
+  void _unregister(int viewId, ContentSensitivity widgetSensitivityLevel) {
+    // Update SensitiveContent widget count for those with
+    // desiredSensitivityLevel.
+    final ViewContentSensitivityState contentSensitivityStateForView = _contentSensitivityStates[viewId]!;
+    contentSensitivityStateForView.removeWidgetWithContentSensitivity(widgetSensitivityLevel);
+
+    if (contentSensitivityStateForView.getTotalNumberOfWidgets() == 0) {
```

Comment:
Maybe I'm getting ahead of myself, but I feel like here as well as in _register above you could use shouldSetContentSensitivity with the previous and next content sensitivity instead of needing to know the total number of widgets. And possibly not need the early return logic where you use numWidgetsWithWidgetSensitivityLevel below or even the `switch` below.